### PR TITLE
kops-controller

### PR DESCRIPTION
**Pull request by justinsb**
kops-controller runs as a container on the master node(s).  It is a kubebuilder
controller, that performs runtime reconciliation for kops.  The first controller is NodeController, which applies labels to nodes.  Previously this was done by the kubelet, but the fear was that this
was insecure and so [this functionality was removed](https://github.com/kubernetes/enhancements/blob/master/keps/sig-auth/0000-20170814-bounding-self-labeling-kubelets.md).

More docs in docs/architecture/kops-controller.md

### CODE REVIEW

File: upup/models/cloudup/resources/addons/kops-controller.addons.k8s.io/k8s-1.16.yaml.template [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+      containers:
+      - name: kops-controller
+        image: kope/kops-controller:1.14.0-alpha.1
+        command:
+{{ range $arg := KopsControllerArgv }}
+        - "{{ $arg }}"
+{{ end }}
+{{- if .EgressProxy }}
+        env:
+{{ range $name, $value := ProxyEnv }}
```
Comment (zetaab) [inline upup/models/cloudup/resources/addons/kops-controller.addons.k8s.io/k8s-1.16.yaml.template]: This is going to need similar env vars what we have in https://github.com/zetaab/kops/blob/1adaf0cceae00f83238a3fc7da24e678bc5362b1/pkg/model/components/etcdmanager/model.go#L462

Comment (zetaab) [reply upup/models/cloudup/resources/addons/kops-controller.addons.k8s.io/k8s-1.16.yaml.template]: so not only if egressproxy is defined, we need another variables as well which is needed for kops state store & cloudprovider things

**Comment by mazzy89** [comment]
Do you need an help here @justinsb? I'd not mind to help to develop the kube-controller.

**Comment by zetaab** [comment]
@justinsb I can implement openstack nodeidentity, just leave me some days before you release 1.16 alpha :)

**Comment by justinsb** [comment]
Cleaned up this PR to be (more) reviewable - and tested it manually against 1.16 on AWS and GCE.  If we can get this in, hopefully we can get our tests working on the k8s 1.16 branch again (or 1.17!).

I might need some help with OpenStack and other clouds, as I don't have easy access to them.

**Comment by rifelpet** [comment]
The pull-kops-e2e-kubernetes-aws test is still pinned to k8s 1.15, so we wont have meaningful prow test results until after this is merged and we unpin k8s in test-infra, correct?

**Comment by justinsb** [comment]
That's right @rifelpet ... I did run the tests manually, but we won't get real results until we unpin.

Some options:

* Accept that I'm committed to making this work if I break anything (we have a 1.15 branch, and this is basically a blocker for 1.16 IMO)
* I think because we are now labelling the masters in protokube that should be sufficient to allow us to test with 1.16 again (we won't label the nodes, but I don't _think_ anything in e2e relies on the nodes).  So we could unpin the tests so they use 1.16 / 1.17, they should pass generally.

In both cases, the risk is that we block other PRs because this PR is broken.  But ... as this PR is passing the current tests, this doesn't seem too likely (at least until we unpin the e2e).

**Comment by rifelpet** [comment]
Ok, thats what I figured. Thanks! I'm looking forward to getting this merged and getting our test grid green again. I'll see if I can test it out myself sometime soon.

File: upup/models/cloudup/resources/addons/kops-controller.addons.k8s.io/k8s-1.16.yaml.template [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+      - key: "node-role.kubernetes.io/master"
+        effect: NoSchedule
+      nodeSelector:
+        node-role.kubernetes.io/master: ""
+      dnsPolicy: Default  # Don't use cluster DNS (we are likely running before kube-dns)
+      hostNetwork: true
+      serviceAccount: kops-controller
+      containers:
+      - name: kops-controller
+        image: kope/kops-controller:1.14.0-alpha.1
```
Comment (zetaab) [inline upup/models/cloudup/resources/addons/kops-controller.addons.k8s.io/k8s-1.16.yaml.template]: 1.14.0-alpha.1? Should we use 1.16.0-alpha.1? If I define `KOPSCONTROLLER_IMAGE` env var this is overrided?

**Pull request merged**